src/core/thread/wakeup_tx_scheduler.hpp

```diff
@@ -0,0 +1,125 @@
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ *  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ *  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ *  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ *  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ *  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ *  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ *  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#ifndef WAKEUP_TX_SCHEDULER_HPP_
+#define WAKEUP_TX_SCHEDULER_HPP_
+
+#include "openthread-core-config.h"
+
+#if OPENTHREAD_CONFIG_WAKEUP_COORDINATOR_ENABLE
+
+#include "common/locator.hpp"
+#include "common/non_copyable.hpp"
+#include "common/timer.hpp"
+#include "mac/mac.hpp"
+
+namespace ot {
+
+class Child;
+
+/**
+ * This class implements wake-up sequence TX scheduling functionality.
+ *
+ */
+class WakeupTxScheduler : public InstanceLocator, private NonCopyable
+{
+    friend class Mac::Mac;
+
+public:
+    /**
+     * Initializes the wake-up sequence TX scheduler object.
+     *
+     * @param[in]  aInstance   A reference to the OpenThread instance.
+     *
+     */
+    explicit WakeupTxScheduler(Instance &aInstance);
+
+    /**
+     * Initiates the wake-up sequence to a Wake-up End Device.
+     *
+     * @param[in] aWedAddress The extended address of the Wake-up End Device.
+     * @param[in] aIntervalUs An interval between consecutive wake-up frames (in microseconds).
+     * @param[in] aDurationMs Duration of the wake-up sequence (in milliseconds).
+     *
+     * @retval kErrorNone         Successfully started the wake-up sequence.
+     * @retval kErrorInvalidState This or another device is currently being woken-up.
+     */
+    Error WakeUp(const Mac::ExtAddress &aWedAddress, uint16_t aIntervalUs, uint16_t aDurationMs);
+
+    /**
+     * Returns the connection window used by this device.
+     *
+     * The connection window is amount of time that this device waits for an initial link establishment message after
+     * sending the last wake-up frame.
+     *
+     * @returns Connection window in the units of microseconds.
+     */
+    uint32_t GetConnectionWindowUs() const { return mIntervalUs * kConnectionRetryInterval * kConnectionRetryCount; }
+
+    /**
+     * Returns the end of the wake-up sequence time.
+     *
+     * @returns End of the wake-up sequence time.
+     */
+    TimeMicro GetTxEndTime() const { return mTxEndTimeUs; }
+
+    /**
+     * Stops the ongoing wake-up sequence.
+     */
+    void Stop(void);
+
+private:
+    using WakeupTimer = TimerMicroIn<Mac::Mac, &Mac::Mac::RequestWakeupFrameTransmission>;
```

Comment:
I know this code works correctly, but it appears a bit unusual that a `Timer` instance defined in the `WakeupTxScheduler` class uses a callback from `Mac::Mac`. 
- The declaration `TimerMicroIn<Mac::Mac, ...>` reads as if this is a `Timer` in the `Mac::Mac` class instead.
- My suggestion is to define a callback method in `WakeupTxScheduler` class for this timer that would then call `Get<Mac::Mac>().RequestWakeupFrameTransmission()`.